File: android/app/src/main/java/app/notomorrow/feature/bro/BroSharedWeekCard.kt
package app.notomorrow.feature.bro

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import app.notomorrow.designsystem.Hairline
import app.notomorrow.designsystem.NT
import app.notomorrow.designsystem.NTCard
import app.notomorrow.designsystem.NtText
import app.notomorrow.designsystem.TabularText
import app.notomorrow.service.DayState
import app.notomorrow.service.WeekDay
import app.notomorrow.util.Fmt
import app.notomorrow.util.S
import app.notomorrow.util.rememberNtStrings
import java.time.Instant
import java.time.ZoneId

/**
 * The one card on the Bro tab — the port of `BroSharedWeekCard`
 * (`Features/Bro/BroSharedWeekCard.swift`): Mon–Sun letters, a row of cells for me and one
 * for the partner, then the next session line with who confirmed when.
 */
@Composable
fun BroSharedWeekCard(
    week: List<WeekDay>,
    partnerName: String,
    session: BroSessionLine?,
    modifier: Modifier = Modifier,
) {
    NTCard(modifier = modifier, padding = 0.dp) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = 14.dp, start = 16.dp, end = 16.dp, bottom = 16.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            LettersRow(week)
            CellsRow(week, stringResource(S.bro_you)) { it.myState }
            CellsRow(week, partnerName) { it.partnerState }
            if (session != null) {
                Hairline(Modifier.padding(top = 4.dp))
                SessionRow(session, partnerName)
            }
        }
    }
}

/** `labelWidth` — the leading name column, 52 pt on both rows. */
private val LabelWidth: Dp = 52.dp

@Composable
private fun LettersRow(week: List<WeekDay>) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Spacer(Modifier.width(LabelWidth))
        Row(Modifier.weight(1f)) {
            week.forEach { day ->
                Cell {
                    NtText(
                        text = stringResource(day.shortLabelRes),
                        style = NT.Fonts.caption,
                        color = if (day.isToday) NT.Colors.ink else NT.Colors.ink2,
                        maxLines = 1,
                    )
                }
            }
        }
    }
}

@Composable
private fun CellsRow(week: List<WeekDay>, label: String, state: (WeekDay) -> DayState) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        NtText(
            text = label,
            modifier = Modifier.width(LabelWidth),
            style = NT.Fonts.footnote,
            color = NT.Colors.ink,
            maxLines = 1,
        )
        Row(Modifier.weight(1f)) {
            week.forEach { day ->
                Cell { BroStatusCell(state = state(day), size = 30.dp) }
            }
        }
    }
}

/** `.frame(maxWidth: .infinity)` on each of the seven columns — equal width, centred. */
@Composable
private fun RowScope.Cell(content: @Composable () -> Unit) {
    Box(Modifier.weight(1f), contentAlignment = Alignment.Center) { content() }
}

@Composable
private fun SessionRow(session: BroSessionLine, partnerName: String) {
    val strings = rememberNtStrings()
    val you = stringResource(S.bro_you)
    val noOneYet = stringResource(S.bro_noOneYet)

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            NtText(
                text = headline(session, strings),
                style = NT.Fonts.headline,
                color = NT.Colors.ink,
                maxLines = 1,
            )
            TabularText(
                text = confirmations(session, partnerName, you, noOneYet),
                style = NT.Fonts.footnote,
                color = NT.Colors.ink2,
            )
        }
        StatusBadge(session, partnerName)
    }
}

@Composable
private fun StatusBadge(session: BroSessionLine, partnerName: String) {
    val text: String
    val dot: androidx.compose.ui.graphics.Color
    when {
        session.bothIn -> {
            text = stringResource(S.bro_bothIn)
            dot = NT.Colors.good
        }
        session.partnerState.isMissedOrCancelled -> {
            text = stringResource(S.bro_today_out_s, partnerName)
            dot = NT.Colors.bad
        }
        session.myState.isMissedOrCancelled -> {
            text = stringResource(S.bro_today_youOut)
            dot = NT.Colors.bad
        }
        else -> return
    }
    Row(
        horizontalArrangement = Arrangement.spacedBy(6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(Modifier.size(8.dp).background(dot, CircleShape))
        NtText(text = text, style = NT.Fonts.footnoteBold, color = NT.Colors.ink, maxLines = 1)
    }
}

/** "Today · 18:00 · Push A" */
private fun headline(session: BroSessionLine, strings: app.notomorrow.util.Localizer): String {
    val parts = mutableListOf(
        Fmt.relativeDay(session.day, strings),
        Fmt.time(session.minuteOfDay),
    )
    session.routineName?.let { parts.add(it) }
    return parts.joinToString(" · ")
}

/** "You 09:12 · Tomek 12:40", or "No one's in yet". */
private fun confirmations(
    session: BroSessionLine,
    partnerName: String,
    you: String,
    noOneYet: String,
    zone: ZoneId = ZoneId.systemDefault(),
): String {
    val parts = mutableListOf<String>()
    session.myConfirmedAt?.let { parts.add("$you ${Fmt.time(Instant.ofEpochMilli(it), zone = zone)}") }
    session.partnerConfirmedAt?.let { parts.add("$partnerName ${Fmt.time(Instant.ofEpochMilli(it), zone = zone)}") }
    return if (parts.isEmpty()) noOneYet else parts.joinToString(" · ")
}

File: NoTomorrow/Features/Bro/BroSharedWeekCard.swift
import SwiftUI

/// The one card on the Bro tab, and its focal card (v2): Mon–Sun letters, a row of cells for me and one for the
/// partner, then the next session line with who confirmed when.
struct BroSharedWeekCard: View {
    var week: [WeekDay]
    var partnerName: String
    var session: BroSessionLine?

    private let labelWidth: CGFloat = 52

    var body: some View {
        FocalCard(padding: 0) {
            VStack(alignment: .leading, spacing: 10) {
                lettersRow
                cellsRow(label: Text("bro.you")) { $0.myState }
                cellsRow(label: Text(partnerName)) { $0.partnerState }
                if let session {
                    Hairline().padding(.top, 4)
                    sessionRow(session)
                }
            }
            .padding(EdgeInsets(top: 14, leading: 16, bottom: 16, trailing: 16))
        }
    }

    private var lettersRow: some View {
        HStack(spacing: 10) {
            Color.clear.frame(width: labelWidth, height: 1)
            HStack(spacing: 0) {
                ForEach(week) { day in
                    Text(LocalizedStringKey(AttendanceService.labelKey(isoWeekday: day.isoWeekday)))
                        .font(NT.Fonts.caption)
                        .foregroundStyle(day.isToday ? NT.Colors.ink : NT.Colors.ink2)
                        .frame(maxWidth: .infinity)
                }
            }
        }
    }

    private func cellsRow(label: Text, state: @escaping (WeekDay) -> DayState) -> some View {
        HStack(spacing: 10) {
            label
                .font(NT.Fonts.footnote)
                .foregroundStyle(NT.Colors.ink)
                .lineLimit(1)
                .frame(width: labelWidth, alignment: .leading)
            HStack(spacing: 0) {
                ForEach(week) { day in
                    BroStatusCell(state: state(day), size: 30)
                        .frame(maxWidth: .infinity)
                }
            }
        }
    }

    private func sessionRow(_ session: BroSessionLine) -> some View {
        HStack(alignment: .center, spacing: 8) {
            VStack(alignment: .leading, spacing: 2) {
                Text(headline(session))
                    .font(NT.Fonts.headline)
                    .foregroundStyle(NT.Colors.ink)
                    .lineLimit(1)
                Text(confirmations(session))
                    .font(NT.Fonts.footnote)
                    .foregroundStyle(NT.Colors.ink2)
                    .tabular()
                    .lineLimit(1)
            }
            Spacer(minLength: 8)
            statusBadge(session)
        }
    }

    @ViewBuilder
    private func statusBadge(_ session: BroSessionLine) -> some View {
        if session.bothIn {
            HStack(spacing: 6) {
                Circle().fill(NT.Colors.good).frame(width: 8, height: 8)
                Text("bro.bothIn").font(NT.Fonts.footnoteBold).foregroundStyle(NT.Colors.ink)
            }
        } else if session.partnerState.isMissedOrCancelled {
            HStack(spacing: 6) {
                Circle().fill(NT.Colors.bad).frame(width: 8, height: 8)
                Text("bro.today.out \(partnerName)").font(NT.Fonts.footnoteBold).foregroundStyle(NT.Colors.ink)
            }
        } else if session.myState.isMissedOrCancelled {
            HStack(spacing: 6) {
                Circle().fill(NT.Colors.bad).frame(width: 8, height: 8)
                Text("bro.today.youOut").font(NT.Fonts.footnoteBold).foregroundStyle(NT.Colors.ink)
            }
        }
    }

    /// "Today · 18:00 · Push A"
    private func headline(_ session: BroSessionLine) -> String {
        var parts = [Fmt.relativeDay(session.day), Fmt.time(minuteOfDay: session.minuteOfDay)]
        if let routine = session.routineName { parts.append(routine) }
        return parts.joined(separator: " · ")
    }

    /// "You 09:12 · Tomek 12:40", or "No one's in yet".
    private func confirmations(_ session: BroSessionLine) -> String {
        var parts: [String] = []
        if let at = session.myConfirmedAt { parts.append("\(String(localized: "bro.you")) \(Fmt.time(at))") }
        if let at = session.partnerConfirmedAt { parts.append("\(partnerName) \(Fmt.time(at))") }
        return parts.isEmpty ? String(localized: "bro.noOneYet") : parts.joined(separator: " · ")
    }
}
